### DISCUSSION TITLE
workflow slowly fills up RAM and crashes

### DISCUSSION THREAD

**Discussion by jbscherrer** [bonsai-rx/bonsai#1394] [Q&A]
Many of my recording workflows will slowly use more and more RAM as they run, usually getting up to about 3 GB of memory after 24 hours of runtime. At this point the latencies become unacceptable, and then at some point certain elements of the workflow hang. It also takes several minutes for the workflow to stop when I issue a stop command. This is an issue for me since I need to be able to run workflows for multiple days at a time.

I've attached an example workflow below that exhibits this behavior, although it's a relatively large workflow (probably riddled with bad bonsai coding practice) and may take a while to inspect completely. I initially suspected that motifs using `WindowTrigger` nodes such as the one below were causing this problem, however I imposed `Count` or `TimeSpan` limits on all of them and it didn't fix the problem.
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/75e50ad4-6284-43b7-acc9-701819626061)

Are there other common bonsai coding mistakes that are known to cause this problem? If necessary I can try to narrow down the culprit to a particular section of the code, though that will take a while since it requires several hours for the problem to become apparent.

[featherscope-bonsai.zip](https://github.com/bonsai-rx/bonsai/files/11616279/featherscope-bonsai.zip)

**Comment by bruno-f-cruz** [comment]
Heya,
I can't really go deep into the workflow because at some point it becomes hard to parse what is going on (As a side note, the readability of your workflow could be substantially improved if you use subjects instead of connecting so many operators with edges.).
At the end of the day, the best solution is to start disabling nodes until the problem disappears. Once you narrow it down, it might even be the case that it is actually a memory leak with some operator and not necessarily a problem with your workflow. 

Anyway, a pattern that jumped out is this Zip at the top of the workflow:
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/ef21aede-e3db-4098-aea8-40ce5756148a)

This is a prime suspect. If your VideoCaptureDevice is not exactly running at 1/T, and is producing slightly more FPS than expected, your "zip" will keep accumulating unpaired frames in memory.

Something else that I noticed is that a lot of SelectMany are not explicitly closed. Most of them should be ok, but i would advise you to be explicit about it and define the end of the sequence just to make sure you don't keep the observables alive. A simple Take(1) into WorkflowOutput should do the trick most of the times.

e.g. 
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/06238a99-072c-4697-9d13-48447247e55c)
Will be alive forever.


For example, i suspect that this SelectMany might never close (I am probably wrong, but just in case give it a check)
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/9b5a4cf6-2b65-4408-a68c-94d10a602c82)

I hope some of these help! good luck!

**Reply by jbscherrer** [reply]
Thank you, these are some great pointers for me to get started! And yeah, no expectation you'd be able to make it through the entire pile of code. Believe it or not, it was even less readable before I recently started trying to improve my use of subjects :P

**Comment by jbscherrer** [comment]
For anyone interested, the dropped frame detector Bruno mentioned can be rewritten without a timer as:
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/b625ce45-3e1e-4885-a9db-f4df7d01205a)

**Comment by jbscherrer** [comment]
As a followup for anyone who may find this useful, I tracked down a devilishly simple part of the workflow that was eating up enormous amounts of RAM, for the reason that Bruno suspected of SelectMany nodes not closing. The script generates a "heartbeat" signal that tells downstream hardware that the script is still running, to prevent hardware from getting stuck in bad states if the computer crashes (eg laser shutter gets held open and bleaches the brain). I was generating this signal using this structure:
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/5db4f05b-06ef-40f6-92ad-3312cfe74ea6)
and within the SelectMany:
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/33d37b35-afc2-4740-a45e-b402ec3f9b27)

Every beat opened a new SelectMany that wasn't closing because the internal sequence had no termination. This was easily fixed by adding a Take(2) node:
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/816bfa23-f915-452d-b9e8-32c5c23bf8b7)
